zerver/actions/streams.py

```diff
@@ -648,18 +649,46 @@ def send_peer_subscriber_events(
                     send_event_on_commit(realm, event, peer_user_ids)
 
         for user_id, stream_ids in user_streams.items():
-            if stream_id in web_public_stream_ids:
-                peer_user_ids = web_public_peer_ids - altered_user_ids
-            else:
-                peer_user_ids = (non_guest_user_ids | subscriber_dict[stream_id]) - altered_user_ids
-
-            event = dict(
-                type="subscription",
-                op=op,
-                stream_ids=sorted(stream_ids),
-                user_ids=[user_id],
-            )
-            send_event_on_commit(realm, event, peer_user_ids)
+            web_public_user_stream_ids = []
+            public_user_stream_ids_without_guest_users = []
+            for stream_id in stream_ids:
+                if stream_id in web_public_stream_ids:
+                    web_public_user_stream_ids.append(stream_id)
+                else:
+                    if (non_guest_user_ids | subscriber_dict[stream_id]) == non_guest_user_ids:
+                        public_user_stream_ids_without_guest_users.append(stream_id)
+                    else:
+                        # If the stream has guest users, we will send
+                        # an event right away.
```

Comment:
Right away seems like the wrong framing here -- I think we should be saying 

"This channel likely has a unique set of peer_user_ids to notify, so we send the event directly for just this channel."